governance/program/src/processor/process_update_program_metadata.rs

```diff
@@ -0,0 +1,67 @@
+//! Program state processor
+
+use borsh::BorshSerialize;
+use solana_program::{
+    account_info::{next_account_info, AccountInfo},
+    clock::Clock,
+    entrypoint::ProgramResult,
+    msg,
+    pubkey::Pubkey,
+    rent::Rent,
+    sysvar::Sysvar,
+};
+use spl_governance_tools::account::create_and_serialize_account_signed;
+
+use crate::state::{
+    enums::GovernanceAccountType,
+    program_metadata::{get_program_metadata_data, get_program_metadata_seeds, ProgramMetadata},
+};
+
+/// Processes UpdateProgramMetadata instruction
+pub fn process_update_program_metadata(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+) -> ProgramResult {
+    let account_info_iter = &mut accounts.iter();
+
+    let program_metadata_info = next_account_info(account_info_iter)?; // 0
+    let payer_info = next_account_info(account_info_iter)?; // 1
+    let system_info = next_account_info(account_info_iter)?; // 2
+
+    let rent = Rent::get().unwrap();
+    let updated_at = Clock::get().unwrap().slot;
+
+    const VERSION: &str = env!("CARGO_PKG_VERSION");
+
+    // Put the metadata info into the logs to make it possible to extract it using Tx simulation
+    msg!("PROGRAM-VERSION:{:?}", VERSION);
+
+    if program_metadata_info.data_is_empty() {
+        let program_metadata_data = ProgramMetadata {
+            account_type: GovernanceAccountType::ProgramMetadata,
+            updated_at,
+            version: VERSION.to_string(),
+            reserved: [0; 64],
+        };
+
+        create_and_serialize_account_signed(
+            payer_info,
+            program_metadata_info,
+            &program_metadata_data,
+            &get_program_metadata_seeds(),
+            program_id,
+            system_info,
+            &rent,
+        )?;
+    } else {
+        let mut program_metadata_data =
+            get_program_metadata_data(program_id, program_metadata_info)?;
```

Comment:
Be sure to validate that `program_metadata_info.key` is the expected metadata address, otherwise this instruction could be maliciously used to clobber the data in any other account type!

Reply:
All the accounts are validated during deserialisation using `get_account_data` and  hence it's not possible to substitute with different account types and the PDA check is not needed  https://github.com/solana-labs/solana-program-library/blob/d6d0b92ae762ac0d4bf726bb28a99cc642e5771b/governance/tools/src/account.rs#L139

Do you think it should check the address anyway?

Reply:
Ah yes you're right, my mistake.  I forgot that `is_initialized()` checks the account type